esmvaltool/diag_scripts/ocean3d/interpolation.py

```diff
@@ -0,0 +1,248 @@
+"""APPLICATE/TRR Ocean Diagnostics."""
+import logging
+import os
+import ESMF
+import numpy as np
+import pyresample
+from cartopy.util import add_cyclic_point
+from netCDF4 import Dataset
+
+from esmvaltool.diag_scripts.ocean3d.getdata import load_meta
+
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+def closest_depth(depths, depth):
+    """Find closest depth.
+
+    From vector of depths finds target depth,
+    that is closest to desired depth, Also returns
+    an index for the desired depth.
+    """
+    target_level = abs(abs(depths) - abs(depth)).argmin()
+    target_depth = depths[target_level]
+    logger.debug('target_depth: %s', target_depth)
+    return target_depth, target_level
+
+
+def interpolate_vert(depth_model, target_depth, data_model):
+    """Vertical linear interpolation.
+
+    Very simple linear interpolation of the model data to the
+    desired depth. Can't extrapolate, so the limitation is that model
+    data should have at least one level => and one level <= than the
+    target depth.
+    """
+    # Simple vertical interpolation
+    dep_up = [z for z in abs(depth_model) if z <= target_depth][-1]
+    dep_lo = [z for z in abs(depth_model) if z > target_depth][0]
+    i_up = 1 - abs(target_depth - dep_up) / (dep_lo - dep_up)
+    i_lo = 1 - abs(target_depth - dep_lo) / (dep_lo - dep_up)
+
+    iz_up = abs(abs(depth_model) - abs(dep_up)).argmin()
+    iz_lo = abs(abs(depth_model) - abs(dep_lo)).argmin()
+
+    data_up = data_model[iz_up, :, :]
+    data_lo = data_model[iz_lo, :, :]
+    if not isinstance(data_up, np.ma.MaskedArray):
+        data_up = np.ma.masked_equal(data_up, 0)
+        data_lo = np.ma.masked_equal(data_lo, 0)
+    data = i_up * data_up
+    data = data + i_lo * data_lo
+    return data
+
+
+def weighting(distance):
+    """Weighting function for pyresample."""
+    weight = 1 / distance**2
+    return weight
+
+
+def interpolate_pyresample(obs_file, mod_file, depth, cmor_var):
+    """The 2d interpolation with pyresample.
```

Comment:
The preprocessor offers 2d regridding functionality. What is the advantage of this function and does it warrant adding another dependency?